include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+ * information as required.
+ */
+struct PluginBase {
+  PluginBase(string_view name, string_view root_id, string_view vm_id,
+             string_view plugin_configuration)
+      : name_(std::string(name)), root_id_(std::string(root_id)), vm_id_(std::string(vm_id)),
+        plugin_configuration_(plugin_configuration) {}
+
+  const std::string name_;
+  const std::string root_id_;
+  const std::string vm_id_;
+  std::string plugin_configuration_;
+  const std::string &log_prefix() const { return log_prefix_; }
+
+private:
+  std::string makeLogPrefix() const;
+
+  std::string log_prefix_;
+};
+
+/**
+ * ContextBase is the interface between the VM host and the VM. It has several uses:
+ *
+ * 1) To provide host-specific implementations of ABI calls out of the VM. For example, a proxy
+ * which wants to provide the ability to make an HTTP call must implement the
+ * ContextBase::httpCall() method.
+ *
+ * 2) To call into the VM. For example, when the above mentioned httpCall() completes, the host must
+ * call ContextBase::onHttpCallResponse(). Similarly, when a new HTTP request arrives and the
+ * headers are available, the host must create a new ContextBase object to manage the new stream and
+ * call onRequestHeaders() on that object which will cause a corresponding Context to be allocated
+ * in the VM which will receive the proxy_on_context_create and proxy_on_request_headers calls.
+ *
+ * 3) For testing and instrumentation the methods of ContextBase can be replaces or augmented.
+ */
+class ContextBase {
+public:
+  ContextBase();                                                   // Testing.
+  ContextBase(WasmBase *wasm);                                     // Vm Context.
+  ContextBase(WasmBase *wasm, std::shared_ptr<PluginBase> plugin); // Root Context.
+  ContextBase(WasmBase *wasm, uint32_t root_context_id,
+              std::shared_ptr<PluginBase> plugin); // Stream context.
+  virtual ~ContextBase();
+
+  WasmBase *wasm() const { return wasm_; }
+  uint32_t id() const { return id_; }
+  bool isVmContext() { return id_ == 0; }
+  bool isRootContext() { return root_context_id_ == 0; }
+  ContextBase *root_context() { return root_context_; }
+  string_view root_id() const { return plugin_->root_id_; }
+  string_view log_prefix() const { return plugin_->log_prefix(); }
+  WasmVm *wasmVm() const;
+
+  //
+  // Calls into the VM.
+  //
+
+  // Root Context
+
+  /**
+   * Call on a Root Context when a VM first starts up.
+   * @param plugin is the plugin which caused the VM to be started.
+   */
+  virtual bool onStart(std::shared_ptr<PluginBase> plugin);
+
+  /**
+   * Call on a Root Context when a plugin is configured or reconfigured dyanmically.
+   * @param plugin is the plugin which is being configured or reconfigured.
+   */
+  virtual bool onConfigure(std::shared_ptr<PluginBase> plugin);
+
+  // Context
+
+  /**
+   * Call on a host Context to create a corresponding Context in the VM.  Note:
```

Comment:
Can you add some color here for implementors? Does an implementation of this live in the SDK or is platform/proxy specific? I'm trying to understand what this does exactly given there is no return value.